sub/sd_sbr.c

```diff
@@ -0,0 +1,188 @@
+        sbr_subtitles_destroy(ctx->sbr_subtitles);
+
+    sbr_subtitle_format fmt = SBR_SUBTITLE_FORMAT_UNKOWN;
+    if(!strcmp(sd->codec->codec, "textsub/srv3") )
+        fmt = SBR_SUBTITLE_FORMAT_SRV3;
+    else if(!strcmp(sd->codec->codec, "textsub/vtt"))
+        fmt = SBR_SUBTITLE_FORMAT_WEBVTT;
+
+    ctx->sbr_subtitles = sbr_load_text(
+        ctx->sbr_library,
+        packet->buffer,
+        packet->len,
+        fmt,
+        NULL
+    );
+    packet->sub_duration = packet->duration;
+}
+
+static struct sub_bitmaps *get_bitmaps(struct sd *sd, struct mp_osd_res dim,
+                                       int format, double pts)
+{
+    struct sd_sbr_priv *ctx = sd->priv;
+    struct mp_subtitle_opts *opts = sd->opts;
+
+    ctx->osd = dim;
+
+    if (pts == MP_NOPTS_VALUE || !ctx->sbr_renderer || !ctx->sbr_subtitles)
+        return NULL;
+
+    if (opts->sub_forced_events_only)
+        return NULL;
+
+    struct sbr_subtitle_context context = (sbr_subtitle_context) {
+        .dpi = 72,
+        .padding_top = (int32_t)dim.mt << 6,
+        .padding_bottom = (int32_t)dim.mb << 6,
+        .padding_left = (int32_t)dim.ml << 6,
+        .padding_right = (int32_t)dim.mr << 6,
+        .video_height = (int32_t)(dim.h - dim.mt - dim.mb) << 6,
+        .video_width = (int32_t)(dim.w - dim.ml - dim.mr) << 6,
+    };
+
+    unsigned t = lrint(pts * 1000);
+
+    struct sub_bitmaps *bitmaps = ctx->bitmaps;
+    struct sub_bitmap *bitmap = bitmaps->parts;
+
+    bool size_did_change = bitmap->w != dim.w || bitmap->h != dim.h;
+    if (size_did_change || sbr_renderer_did_change(ctx->sbr_renderer, &context, t)) {
+        if(bitmaps->packed)
+          talloc_free(bitmaps->packed);
+
+        bitmaps->packed = mp_image_alloc(IMGFMT_BGRA, dim.w, dim.h);
+        bitmaps->packed_h = dim.h;
+        bitmaps->packed_w = dim.w;
+        bitmaps->packed->params.repr.alpha = PL_ALPHA_PREMULTIPLIED;
+        ++bitmaps->change_id;
+
+        bitmap->bitmap = bitmaps->packed->planes[0];
+        bitmap->w = dim.w;
+        bitmap->h = dim.h;
+        bitmap->dw = dim.w;
+        bitmap->dh = dim.h;
+        bitmap->stride = (*bitmaps->packed).stride[0];
+
+        sbr_renderer_render(ctx->sbr_renderer, &context, ctx->sbr_subtitles, t, bitmap->bitmap,
+                            dim.w, dim.h, bitmap->stride >> 2);
+    }
+
+    return sub_bitmaps_copy(NULL, bitmaps);
+}
+
+static void uninit(struct sd *sd)
+{
+    struct sd_sbr_priv *ctx = sd->priv;
+
+    if(ctx->bitmaps->packed)
+        talloc_free(ctx->bitmaps->packed);
+    if(ctx->sbr_renderer)
+        sbr_renderer_destroy(ctx->sbr_renderer);
+    if(ctx->sbr_subtitles)
+        sbr_subtitles_destroy(ctx->sbr_subtitles);
+    sbr_library_fini(ctx->sbr_library);
+}
+
+static int control(struct sd *sd, enum sd_ctrl cmd, void *arg)
+{
+    struct sd_sbr_priv *ctx = sd->priv;
+    switch (cmd) {
+    case SD_CTRL_SET_VIDEO_PARAMS:
+        ctx->video_params = *(struct mp_image_params *)arg;
```

Comment:
Is this used anywhere?

Reply:
Ah, you're right it isn't. Thought it might've been necessary to get the padding but then found it's there in `mp_osd_res`. I'll remove it.